Rework unit reanimation system

cogito123 opened the pull request:
### Brief description of the changes:
- Reorganize unit reanimation logic, so it's clear what is happening;
- Make the system configurable by `global.max_reanim_thresh` parameter to enable reanimation threshold scaling;
- Store and "schedule" biter re-spawn by distributing requests every tick;
- When unit is re-spawned, try to reuse biter corpses on battlefield otherwise fallback and randomize biter unit based on evolution factor;
- FPS / UPS drop of this solution is not noticeable even at 100% reanimation chance and shooting nukes at large hordes;
- Added adjustable reanimation balancer which scales amount of served requests per tick. This helps keep memory low and re-adjust to quickly revive critical mass of biters in case defenses are overpowered.

### Tested Changes:
- [x] Changes tested

AwesomePatrol commented:
I like added event filters, clarified reanimation thresholds, and distribution of clone events across game ticks.
However, I don't understand the reasoning behind the logic change. What is the problem you try to solve? Why is the current implementation bad? It adds yet another global array that is known to cause performance issues due to pointer chasing and GC.

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +480,117 @@ function Public.subtract_threat(entity)
+	'behemoth-biter',
+}
+
+local function likely_biter_name(force_name)
```
I see two main problems with this:
* it looks duplicated (similar function is in a different module),
* it uses current evolution (what isn't a significant difference from the current reanimate logic, but may be not intuitive).

You can simplify this part of the code, by limiting revives to behemot biters/spitters only. This code should only be active after 100% evo, so there is little impact on the gameplay.

cogito123 replied:
> * t looks duplicated (similar function is in a different module),

I think you meant "biter_raffle"? I don't want to use it, since exported API is not compatible with what I'm doing here and I don't want to waste time "rolling" single biter type 50 times like in casino.

> * it uses current evolution (what isn't a significant difference from the current reanimate logic, but may be not intuitive).

Logic before my change also used evolution, it just was hidden by 3 layers of code.

> You can simplify this part of the code, by limiting revives to behemot biters/spitters only. This code should only be active after 100% evo, so there is little impact on the gameplay.

It's forward compatible. This function in my opinion is "complete" as it will work with any evolution factor.

AwesomePatrol replied:
> I don't want to waste time "rolling" single biter type 50 times like in casino.

I agree, so please consider changing it there.

cogito123 commented:
> However, I don't understand the reasoning behind the logic change. What is the problem you try to solve?

During my benchmarking of late game performance I've notice that if multiple biters die at the same tick, we try to revive them all at once within the same tick. It causes many API calls which are expensive. I wanted to distribute the load to other ticks, so we don't suffer from performance degradation at higher evo%.

> Why is the current implementation bad?

There are too many API calls at higher evo% (`LuaSurface::create_entity`). The current code is also obscure to understand so I rewritten it in cohesive way and added corpse re-use so it looks more realistic.

> It adds yet another global array that is known to cause performance issues due to pointer chasing and GC.

Most expensive are API calls. If you create a unit entity, memory has to be allocated, it has to be drawn or screen, perhaps some more initialization logic is executed and then short pathing request is performed. If you call this many times within same tick it will cause lag spike - if this is a prolonged state then game will slow down.

I also don't agree with your concern about performance issue coming from global variable use. Can you point me to source of that information? What GC has to do with it? We don't leave garbage in memory when using global. My assumption is that only local variable are cleared. I made a benchmark and here is the result:
```
Max: 0.8819ms
Min: 0.2633ms
Avg: 0.6709ms
```
This is 1000 iterations of inserting element into global array, accessing it and removing it within same tick. So if you divide average time by 1000, you get effective 6us for this operation on my machine. From my standpoint cost of "global" variable use is non-existent.
```
Max: 0.1172ms
Min: 0.0294ms
Avg: 0.0415ms
```
Accessing only global variable for 1000 iterations, so time for single access in 41.5ns on average.

Code used for first benchmark:
```lua
local profile = game.create_profiler(false)
profile.reset()
for i = 1, 1000, 1 do
	global.dummy[#global.dummy + 1] = i
	local access = global.dummy[1]
	table.remove(global.dummy, 1)
end
profile.stop()
log(profile)
```

Code used for second benchmark:
```lua
local profile = game.create_profiler(false)
profile.reset()
for i = 1, 1000, 1 do
	local access = global.dummy[1]
end
profile.stop()
log(profile)
```

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +481,143 @@ function Public.subtract_threat(entity)
+
+local function likely_biter_name(force_name)
+	-- Get most likely biter name based on current evolution.
+	local count = #UNIT_NAMES
```
Unless you want to modify `UNIT_NAMES`, a static value can be taken out of this function.

AwesomePatrol replied:
Yes, but `count` isn't. It is a nitpick, so I am resolving this.

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +481,143 @@ function Public.subtract_threat(entity)
+
+	-- Find corpse to spawn unit on top of.
+	local surface = game.surfaces[global.bb_surface_name]
+	local corpse = surface.find_entities_filtered {
```
Personally I don't like it. I would just use `likely_biter_name` method as API calls are slow. However, if you intend to keep it, please somehow restrict the impact of `clear-corpses` command.

AwesomePatrol replied:
`likely_biter_name` doesn't correctly match the distribution of biters in a wave. Therefore, you should use biter's corpses as often as possible and `likely_biter_name` only as a fallback. It shouldn't be influenced by player's actions. I image you can time `clear-corpses` command to have lower-tier biters respawn.

There are at least two possible solutions to this:
* restrict timing of the command (can be run only when no biters are attacking -- no biter deaths in last X ticks),
* modify command to remove only old biters.

cogito123 replied:
I choose to restrict timing of the `clear-corpses` command. It will now check if all FIFOs are empty before call is allowed to execute.

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +481,143 @@ function Public.subtract_threat(entity)
+	-- Bother calculating threshold only for evolution less than 100.
+	if evo < 1 then
+		-- Map evolution onto array indicies.
+		idx = math.ceil((evo + 0.1) * count)
```
Does it mean a new biter type every 10% evo? Then the order of the array should be switched - biters usually come before spitters.

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +481,143 @@ function Public.subtract_threat(entity)
+	if idx > 1 then
+		idx = math.random(idx - 1, idx)
+	else
+		idx = math.random(idx, idx + 1)
```
Just don't randomize in this case?

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +481,143 @@ function Public.subtract_threat(entity)
-	if not cut_off then return end
-	reanimate(entity, cut_off)
+	-- Store only position, that is enough to guess force and type of biter.
+	global.dead_units[idx] = fifo.push(global.dead_units[idx], position)
```
It is enough to guess biter type, but if you pushed it to fifo, you would greatly reduce complexity of this code (no need to search for corpse or have `likely_biter_name` function).

AwesomePatrol replied:
You don't need to push the entire `entity`. I would guess that `unit` types are constants / static string references and shouldn't change. Alternatively you may just store an index of type of biter as it definitely won't be affected.

cogito123 replied:
After restricting the `clear-corpses` this is not an issue anymore. There however can still be data race between game engine and revive logic. In case our script works too fast and there are many units dying at the same moment, we'll still be unable to find corpses as they're not spawned yet. In my opinion fallback to `likely_biter_name` is not noticeable.

AwesomePatrol commented on maps/biter_battles_v2/ai.lua:
```diff
@@ -480,49 +481,143 @@ function Public.subtract_threat(entity)
-	group.add_member(unit)
-	entity.destroy()
-end
+	local force = game.forces['south_biters']
```
Why do you want to avoid using `force` from `event`?

AwesomePatrol commented on maps/biter_battles_v2/init.lua:
```diff
@@ -154,12 +154,36 @@ function Public.tables()
+
+	-- Maximum evolution threshold after which biters have 100% chance
+	-- to reanimate. The real evolution value in this case must be
+	-- subtracted by 100, since reanimation starts at that point.
```
I know you mean `real_evo - 100 = max_reanim_thresh`, but it is hard to understand. Please rephrase or add an example

cogito123 replied:
Done.

cogito123 commented on maps/biter_battles_v2/fifo_list.lua:
```diff
@@ -0,0 +1,40 @@
+end
+
+function mod.pop(list)
+	local node = list.next
```
I made FIFO without list nodes and it's kept in global pool memory. There can be still garbage collector involved, but I cannot do that without specializing FIFO to just single type of work. In truly GC-avoidant case I would need all keys to be predefined for every element within init stage of a game. I also want pool memory to be allocator backend and FIFO as container frontend, so it's not all mixed up and can be reused elsewhere.

I'm looking for some other solutions, e.g. write some schema builder, but for now current solution is sufficient.

AwesomePatrol reviewed (approved):
Tested. Works well. In my experiments, `lua garbage incremental` was always below `3.5ms` (usually `0.01ms`) and `script update` below `4.5ms`.

Thanks for this great work.
